common/channel/src/libra_channel.rs

```diff
@@ -42,24 +43,48 @@ struct SharedState<T> {
     /// consumption by the Receiver and then the executor/scheduler will wake up
     /// the Receiver task to process the next item.
     waker: Option<Waker>,
+
+    /// A boolean which tracks whether the receiver has dropped
+    receiver_dropped: bool,
+    /// A boolean which tracks whether the sender has dropped
+    sender_dropped: bool,
+    /// A boolean which tracks whether the stream has terminated
+    /// A stream is considered terminated when sender has dropped
+    /// and we have drained everything inside our internal queue
+    stream_terminated: bool,
 }
 
 /// The sending end of the libra_channel.
-#[derive(Clone)]
 pub struct Sender<T> {
     shared_state: Arc<Mutex<SharedState<T>>>,
 }
 
 impl<T: MessageQueue> Sender<T> {
-    /// This adds the message into the internal queue data structure. This is a non-blocking
+    /// This adds the message into the internal queue data structure. This is a
     /// synchronous call.
     /// TODO: We can have this return a boolean if the queue of a validator is capacity
-    pub fn put(&mut self, key: T::Key, message: T::Message) {
+    pub fn put(&mut self, key: T::Key, message: T::Message) -> Result<()> {
         let mut shared_state = self.shared_state.lock().unwrap();
+        if shared_state.receiver_dropped {
+            return Err(format_err!(
+                "The receiver for this channel has dropped. This should never happen."
```

Comment:
I don't really know about the "This should never happen" part. That's a pretty common thing to occur when using channels.

Reply:
Updated